File: secure-gateway-studio/backend/src/sgstudio/domain/naming.py
from __future__ import annotations

import hashlib


def service_account_id(deployment_name: str, role: str) -> str:
    """Return a stable Google service-account ID within the 30-char limit."""
    candidate = f"{deployment_name}-{role}"
    if len(candidate) <= 30:
        return candidate
    digest = hashlib.sha256(deployment_name.encode()).hexdigest()[:6]
    prefix_length = 30 - len(role) - len(digest) - 2
    prefix = deployment_name[:prefix_length].rstrip("-") or "sgs"
    return f"{prefix}-{digest}-{role}"


def service_account_email(deployment_name: str, project_id: str, role: str) -> str:
    return f"{service_account_id(deployment_name, role)}@{project_id}.iam.gserviceaccount.com"

File: secure-gateway-studio/extension/src/domain/naming.ts
/**
 * Resource naming. Port of `domain/naming.py`.
 *
 * Service-account IDs are capped at 30 characters by Google. When the natural
 * name would overflow, a digest of the deployment name is folded in so the
 * result stays stable for a given deployment rather than being truncated into
 * a collision with a differently-named one.
 */

import { sha256HexOfString } from "./sha256.ts";

/** A stable Google service-account ID within the 30-character limit. */
export function serviceAccountId(deploymentName: string, role: string): string {
  const candidate = `${deploymentName}-${role}`;
  if (candidate.length <= 30) return candidate;
  const digest = sha256HexOfString(deploymentName).slice(0, 6);
  const prefixLength = 30 - role.length - digest.length - 2;
  const prefix = deploymentName.slice(0, prefixLength).replace(/-+$/, "") || "sgs";
  return `${prefix}-${digest}-${role}`;
}

export function serviceAccountEmail(
  deploymentName: string,
  projectId: string,
  role: string,
): string {
  return `${serviceAccountId(deploymentName, role)}@${projectId}.iam.gserviceaccount.com`;
}
